Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -30,167 +30,320 @@
-}
+        id ret = [NSMutableArray array];
 
-static id _mainQueue;
+        if (_curOperation != nil) {
+            [ret addObject:_curOperation];
+        }
 
-@interface NSOperationQueue () {
-    struct NSOperationQueuePriv* priv;
-    dispatch_queue_t _completionQueue;
-}
-@end
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            if (!_operationsHoldingArea[i].empty()) {
+                for (const auto& op : _operationsHoldingArea[i]) {
+                    [ret addObject:op];
+                }
+            }
 
-@implementation NSOperationQueue
-static void ClearList(NSAtomicListRef* listPtr) {
-    for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
-        while (NSAtomicListPop(&listPtr[i])) {
-            ;
+            if (!_candidateOperationsToRunNext[i].empty()) {
+                for (const auto& op : _candidateOperationsToRunNext[i]) {
+                    [ret addObject:op];
+                }
+            }
         }
+
+        return ret;
     }
-}
 
-static BOOL RunOperationFromLists(NSAtomicListRef* listPtr, NSAtomicListRef* sourceListPtr, id* curOperation) {
-    StrongId<NSOperation> op = NSAtomicListPop(listPtr);
-    if (op == nil) {
-        *listPtr = NSAtomicListSteal(sourceListPtr);
-        // source lists are in LIFO order, but we want to execute operations in the order they were enqueued
-        // so we reverse the list before we do anything with it
-        NSAtomicListReverse(listPtr);
-        op = NSAtomicListPop(listPtr);
+    unsigned int GetOperationCount() {
+        std::lock_guard<std::mutex> lock(_lock);
+
+        unsigned int ret = 0;
+
+        if (_curOperation != nil) {
+            ret++;
+        }
+
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            ret += _operationsHoldingArea[i].size();
+            ret += _candidateOperationsToRunNext[i].size();
+        }
+
+        return ret;
+    }
+
+    void CancelAllOperations() {
+        std::lock_guard<std::mutex> lock(_lock);
+
+        if (_curOperation != nil) {
+            [_curOperation cancel];
+        }
+
+        for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
+            if (!_operationsHoldingArea[i].empty()) {
+                for (const auto& op : _operationsHoldingArea[i]) {
+                    [op cancel];
+                }
+            }
+
+            if (!_candidateOperationsToRunNext[i].empty()) {
+                for (const auto& op : _candidateOperationsToRunNext[i]) {
+                    [op cancel];
+                }
+            }
+        }
+    }
+
+    void Resume() {
+        BOOL shouldBroadcast = FALSE;
+        {
+            std::lock_guard<std::mutex> lock(_lock);
+            if (_isSuspended) {
+                _isSuspended = FALSE;
+                shouldBroadcast = TRUE;
+            }
+        }
+
+        if (shouldBroadcast) {
+            _suspendedCondition.notify_all();
```

Comment:
documentation states that notify calls are most efficient outside of the lock. Otherwise it causes a "hurry up and wait" scenario where the notified thread immediately waits again to re-acquire the mutex. Some implementations optimize for this but to be most portable, the recommendation is to notify outside of the lock.

---
In reply to: [95448692](https://github.com/Microsoft/WinObjC/pull/1676#discussion_r95448692) [](ancestors = 95448692)